OriginProtocol/origin-dollar
# Sparrow dom/smoke tests

This PR creates a framework for executing smoke tests. To achieve this: 
- a node:fork is ran and the script waits for it to be ready 
- then all tests are ran from the `scripts/test/smokeTest` folder. Each test has before & after deploy functions
- once all before deploy tests execute the contract deploy script is ran. Specified by the `deployid` parameter or if left blank, the last deploy script is used.
  - A small change is needed in deployments file, where hardhat RE is passed to the main function. See `011_ousd_fix.js` for an example
- Once the contract deploy is done all the after deploy functions execute and perform required checks

Example how to run smoke tests: 
`scripts/test/smokeTest.sh --deployid 11`

Ideas for improvement: 
- Make it integratable with CI: 
  -  fetch one of the latest block numbers and run hardhat node with that env variable
  - use `deployments/mainnet/.migrations` (thanks for the idea @franckc) to figure out which contract deployments haven't yet been ran. And run smoke tests against those.


Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)
